bee-tangle/src/urts.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{ms_tangle::MsTangle, storage::StorageBackend};
+
+use bee_message::MessageId;
+
+use hashbrown::{hash_map::Entry, HashMap, HashSet};
+use log::debug;
+use rand::seq::IteratorRandom;
+
+use std::time::Instant;
+
+enum Score {
+    NonLazy,
+    SemiLazy,
+    Lazy,
+}
+
+// C1: the maximum allowed delta value for the YMRSI of a given message in relation to the current SMI before it
+// gets lazy.
+const YMRSI_DELTA: u32 = 8;
+// C2: the maximum allowed delta value between OMRSI of a given message in relation to the current SMI before it
+// gets semi-lazy.
+const OMRSI_DELTA: u32 = 13;
+// M: the maximum allowed delta value between OMRSI of a given message in relation to the current SMI before it
+// gets lazy.
+/// The maximum permitted difference between a message's OMRSI and the current SMI before it is considered lazy.
+pub const BELOW_MAX_DEPTH: u32 = 15;
+// If the amount of non-lazy tips exceed this limit, remove the parent(s) of the inserted tip to compensate for the
+// excess. This rule helps to reduce the amount of tips in the network.
+const MAX_LIMIT_NON_LAZY: u8 = 100;
+// The maximum time a tip remains in the tip pool after having the first child.
+// This rule helps to widen the tangle.
+const MAX_AGE_SECONDS_AFTER_FIRST_CHILD: u8 = 3;
+// The maximum amount of children a tip is allowed to have before the tip is removed from the tip pool. This rule is
+// used to widen the cone of the tangle.
+const MAX_NUM_CHILDREN: u8 = 2;
+
+#[derive(Default)]
+struct TipMetadata {
+    children: HashSet<MessageId>,
+    time_first_child: Option<Instant>,
+}
+
+impl TipMetadata {
+    pub(crate) fn new() -> Self {
+        Self::default()
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct UrtsTipPool {
+    tips: HashMap<MessageId, TipMetadata>,
+    non_lazy_tips: HashSet<MessageId>,
+}
+
+impl UrtsTipPool {
+    pub(crate) fn non_lazy_tips(&self) -> &HashSet<MessageId> {
+        &self.non_lazy_tips
+    }
+
+    pub(crate) async fn insert<B: StorageBackend>(
+        &mut self,
+        tangle: &MsTangle<B>,
+        message_id: MessageId,
+        parents: Vec<MessageId>,
+    ) {
+        if let Score::NonLazy = self.tip_score::<B>(tangle, &message_id).await {
+            self.non_lazy_tips.insert(message_id);
+            self.tips.insert(message_id, TipMetadata::new());
+            for parent in parents.iter() {
```

Comment:
why not just `&parents`? :trollface:

Reply:
This aside, I have a list of things I think need to change about `Parents`... but that's for another day